### DISCUSSION TITLE
Does cilium/ebpf help with keeping constants and types in sync between Go and C

### DISCUSSION THREAD

**Discussion by christian-2** [cilium/ebpf#1342] [Q&A] [answered]
Does `cilium/ebpf` provide any support for keeping in sync constants and types that refer to maps shared between user-level code written in Go and kernel-level code written in C, or is this out of the scope of this project. I assume it is, but would like to know for sure.

**Comment by lmb** [comment]
Can you give an example what you mean?

**Comment by christian-2** [comment]
I'm fairly new to eBPF programming and this is what I face:

An eBPF program written in C and a user-space program written in Go. Both exchange data via an eBPF map. Conceptually, the eBPF map contains a union, so there is a field of type integer (`__u64` in C, `uint64` in Go) inside that discerns the particular cases. I have a `map.go` with`const { CASE_1 = atop ... }` and a `map.h` that declares `#define CASE_1 0 ...`. 

If the two files get out of sync, things obviously go awry. In order to mitigate against one could link C into Go (this has downsides) or envisage e.g. a tool that can generate one representation from the other (e.g. Go from C).

So this is the area I had in mind. Is it clearer now?

**Reply by brycekahle** [reply, marked as answer]
You should look into using `go tool cgo -godefs types.go` against a file like this:
```
//go:build ignore

package mypkg

/*
#include "./types.h"
*/
import "C"

type MyGoType C.my_c_type
```

**Reply by thediveo** [reply]
how does this work with defines? it seems to cover only types.

**Comment by christian-2** [comment]
@brycekahle Perfect, thank you.
